Introducing new EnumType

ossinkine opened the pull request:
The goal of this PR is an ability to use common form type for enum fields. Also it simplify generated form type classes.
Someone may be embarrassed by using generated classes, which don't placed in CVS and appear only when Symfony container is compiled, also these classes can't be loaded by Composer autoloader (they can be loaded only on Symfony boot). For example, static analyzers can't find the classes and throw alert.

PS: Have you thought about about storing generated classes in project src directory?

ossinkine committed:
Introducing new EnumType

ossinkine commented:
While I was preparing this project, I began to doubt the work done 😄

magnusnordlander commented:
Ping @morticue.

As for storing generated classes in the src directory, I'm generally against it, but could be convinced to make that a configurable option. Re-generating the classes would then be done manually with a command.

ossinkine commented:
@magnusnordlander Why might re-generating be necessary? One-time generating should be enough.

magnusnordlander commented:
Version updates, namespace moves, etc. Some of them might overwrite the contents of the class, some of them might rename files/classes.

If we had a command to generate the classes manually, we'd still want to just run that command once for all enums, and then you'd re-run that command if you add new enums.

morticue commented:
This PR doesn't seem to work for me. I get the following error

```
The form's view data is expected to be an instance of class AppBundle\Entity\Enum\Gender, but is a(n) string. You can avoid this error by setting the "data_class" option to null or by adding a view transformer that transforms a(n) string to an instance of AppBundle\Entity\Enum\Gender.
```

I don't have anything against what this PR tries to achieve. Usability wise I think it could be cleaner to have this in your FormTypes (as this PR suggest):

```
->add('gender', EnumType::class, [
    'data_class' => Gender::class,
])
````

Compared to this:

```
->add('gender', GenderType::class)
```

---

I am also positive to the idea of manually generating the form type classes from a command which create classes under the `src/` directory. 
The above (old) "automatic" method could still be there, and be the default behaviour of the bundle. But can be disabled via config. So you have to create the form types yourself, if you need them, by running the command.

ossinkine commented:
@morticue This does not work for you, because your root object returns string for `gender` field, for example `Gender::MALE` instead of `Gender::MALE()` which returns an object.

morticue commented:
I don't think that is it. In a formtype where I experimented with this PR, I have a field that looks like this (I followed what you wrote in the readme):

```
->add('status', \Fervo\EnumBundle\Form\EnumType::class, [
    'data_class' => \AppBundle\Enum\PostStatus::class,
    'data' => $options['article']->getStatus(),
    'label' => 'Status',
])
```

The underlying data for the field is passed in the `data`-attribute with `$options['article']->getStatus()` which contains this:

```
object(AppBundle\Enum\PostStatus)[1017]
  protected 'value' => string 'published' (length=9)
```

That is an object (the enum), and not the string value of the enum. But I get this error:

```
The form's view data is expected to be an instance of class AppBundle\Enum\PostStatus, but is a(n) string. You can avoid this error by setting the "data_class" option to null or by adding a view transformer that transforms a(n) string to an instance of AppBundle\Enum\PostStatus.
```

ossinkine commented:
@morticue Could you please publish the reproducer?

morticue commented:
@ossinkine I have created a git repo on the url below, containing a small app which reproduce the error I encounter:

https://github.com/morticue/FervoEnumBundlePR7

```
$ git clone https://github.com/morticue/FervoEnumBundlePR7.git
$ cd FervoEnumBundlePR7/
$ composer install --no-interaction
$ bin/console server:run
```

Browse to http://127.0.0.1:8000/ and the following error occurs:

```
The form's view data is expected to be an instance of class AppBundle\Enum\Gender, but is a(n) string. You can avoid this error by setting the "data_class" option to null or by adding a view transformer that transforms a(n) string to an instance of AppBundle\Enum\Gender.
```

ossinkine commented:
@morticue Thank you for report. I've updated the PR, please review. I've changed `data_class` option to `class` (like in `Symfony\Bridge\Doctrine\Form\Type\EntityType`).

morticue commented:
@ossinkine Nice! Now the form type works 🙂 

One other thing though. In your PR the functionality around the translation of the form label has changed with the `choice_label_prefix` option, and its sensible default where it prefixes the translation key with the class name, snake_case style.

Previously the translation prefix in the form type was defined in config, and the config for it is still there in the bundle config in this PR, but it is no longer used (in the form type).

That in itself could be a way forward, especially since your PR is a BC break regardless.

However, another part of the bundle still uses that config, and that is the twig extension `enum_trans`
https://github.com/ossinkine/EnumBundle/blob/master/Twig/EnumExtension.php#L34

This means that an enum could have multiple translations depending on if it is used in form types and if it’s used in the twig extension. This seems less than ideal.

Do you have any thoughts on a mitigation strategy?

ossinkine commented:
Translations from config are still used in generated form types, so actually there is no BC breaks.
I've just introduced a new implementation of enum form type which can be used by itself. If you update this bundle in your project, you don't see differences.

morticue commented:
Ah, my bad. You are right! 👍 

I will merge this now, and tag a new release 2.1.0. Thanks!
